pkg/recovery/apiserver.go

```diff
@@ -0,0 +1,393 @@
+package recovery
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"text/template"
+	"time"
+
+	"github.com/ghodss/yaml"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	clientcmdapiv1 "k8s.io/client-go/tools/clientcmd/api/v1"
+	"k8s.io/klog"
+
+	"github.com/openshift/library-go/pkg/crypto"
+
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/v311_00_assets"
+)
+
+const (
+	KubeApiserverStaticPodFileName = "kube-apiserver-pod.yaml"
+	RecoveryPodFileName            = "recovery-kube-apiserver-pod.yaml"
+	RecoveryPodNamespace           = "kube-system"
+	RecoveryCofigFileName          = "config.yaml"
+	AdminKubeconfigFileName        = "admin.kubeconfig"
+
+	RecoveryPodAsset    = "v3.11.0/kube-apiserver/recovery-pod.yaml"
+	RecoveryConfigAsset = "v3.11.0/kube-apiserver/recovery-config.yaml"
+)
+
+type Apiserver struct {
+	KubeApiserverImage    string
+	PodManifestDir        string
+	ResourceDirPath       string
+	StaticPodResourcesDir string
+	recoveryDir           string
+
+	kubeApiserverStaticPod *corev1.Pod
+	restConfig             *rest.Config
+	kubeClientSet          *kubernetes.Clientset
+}
+
+// TODO: rename manifest -> pod
+func (s *Apiserver) GetCurrentApiserverManifest() (*corev1.Pod, error) {
+	if s.kubeApiserverStaticPod != nil {
+		return s.kubeApiserverStaticPod, nil
+	}
+
+	kubeApiserverManifestPath := path.Join(s.PodManifestDir, KubeApiserverStaticPodFileName)
+	pod, err := ReadManifestToV1Pod(kubeApiserverManifestPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode file %q: %v", kubeApiserverManifestPath, err)
+	}
+
+	s.kubeApiserverStaticPod = pod
+
+	return s.kubeApiserverStaticPod, nil
+}
+
+func (s *Apiserver) GetResourceDirPath() (string, error) {
+	if s.ResourceDirPath != "" {
+		return s.ResourceDirPath, nil
+	}
+
+	pod, err := s.GetCurrentApiserverManifest()
+	if err != nil {
+		return "", fmt.Errorf("failed to read apiserver pod manifest: %v", err)
+	}
+
+	resourceDir, err := GetVolumeHostPathPath("resource-dir", pod.Spec.Volumes)
+	if err != nil {
+		return "", fmt.Errorf("failed to find resource-dir: %v", err)
+	}
+
+	s.ResourceDirPath = *resourceDir
+
+	return s.ResourceDirPath, nil
+}
+
+func (s *Apiserver) GetRecoveryDir() string {
+	return s.recoveryDir
+}
+
+func (s *Apiserver) RestConfig() (*rest.Config, error) {
+	if s.restConfig == nil {
+		return nil, errors.New("no rest config is set yet")
+	}
+
+	return s.restConfig, nil
+}
+
+func (s *Apiserver) KubeConfig() (*clientcmdapiv1.Config, error) {
+	restConfig, err := s.RestConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	return &clientcmdapiv1.Config{
```

Comment:
`rest.CopyConfig`

Reply:
from where? you are also commenting on the wrong PR, this one is just picking up https://github.com/openshift/cluster-kube-apiserver-operator/pull/448